advanced_source/extend_dispatcher.rst

```diff
@@ -0,0 +1,295 @@
+Extending dispatcher for a new backend in C++
+=============================================
+
+In this tutorial we will walk through all necessary steps to extend the dispatcher to
+add a new backend living outside ``pytorch/pytorch`` repo and maintain it to keep in
+sync with native PyTorch backends.  Here we'll assume that you're familiar with how
+to `register a dispatched operator in C++ <dispatcher>`_ and how to write a
+`custom autograd function <cpp_autograd>`_.
+
+
+.. note::
+
+   This tutorial touches a lot of internal components inside PyTorch which are being actively improved,
+   please expect changes to APIs if you decide to follow this tutorial.  We'll keep this tutorial
+   up to date with the latest APIs.
+
+What's a valid new backend?
+---------------------------
+
+Before adding a new backend, let's first look at your use cases.  If you have better ideas about
+how to make a few operators run faster on CPU/GPU, you don't need create a new backend
+to enable the optimization.  You are welcome to send PRs to ``pytorch/pytorch`` to
+upstream those CPU/CUDA kernels directly.  Adding a new backend is only applicable for cases that
+the way you write kernels are fundamentally different, for example:
+
+* supporting a new hardware like Google TPU and customized chips, which often requires using
+  hardware-specific API to write kernels.
+* supporting operators with a different Tensor layout/representation like sparse and quantized, which
+  enforces your kernels to be written in a way that's more efficient given the layout/representation
+  limitation.
+
```

Comment:
I'd probably want a pointer to how to setup build system for C++ extension to PyTorch somewhere. Fortunately we have preexisting infrastructure so you don't have to redux it here.